src/jmatrix.jl

```diff
@@ -1,110 +1,108 @@
 """
-JMatrix(T::Type, n::Integer) = JMatrix{T,typeof(n),Any}(n, T(1))
-JMatrix(T::Type, n::Integer, sgn) = JMatrix{T,typeof(n), Any}(n, T(sgn))
-
-Base.similar(J::JMatrix,::Type{T}) where {T} = JMatrix(T, J.n, J.sgn)
-Base.similar(J::JMatrix, ::Type{T}, dims::Union{Dims{1},Dims{2}}) where {T} = zeros(T, dims...)
-Base.copy(J::JMatrix{T}) where T = JMatrix(T, J.n, J.sgn)
+JMatrix{T, ±1}(n)
 
-Base.size(J::JMatrix) = (J.n, J.n)
-function Base.size(J::JMatrix, n::Integer)
-    if n == 1 || n == 2
-        return J.n
-    else 
-        return 1
+Creates an `AbstractMatrix{T}` of size `n x n`, representing a
+block-diagonal matrix whose diagonal blocks are `±[0 1; -1 0]`.
+If `n` is odd, then the last block is the `1 x 1` zero block.
+The `±1` parameter allows us to transpose and invert the matrix,
+and corresponds to an overall multiplicative sign factor.
+"""
+struct JMatrix{T<:Real, SGN} <: AbstractMatrix{T}
+    n::Int    # size of the square matrix
+    function JMatrix{T, SGN}(n::Integer) where {T, SGN}
+        n ≥ 0 || throw("size $n must be ≥ 0")
+        (SGN === +1 || SGN === -1) || throw("SGN parameter must be ±1")
+        new{T, SGN}(n)
     end
 end
 
-function Base.Matrix(J::JMatrix{T}) where {T}
-    M = similar(J, T, J.n, J.n)
-    for i =1 : 2 : J.n-1
-        M[i+1,i] = -J.sgn
-        M[i,i+1] = J.sgn
+Base.size(J::JMatrix) = (J.n, J.n)
+Base.size(J::JMatrix, n::Integer) = n in (1,2) ? J.n : 1
+
+function Base.Matrix(J::JMatrix{T, SGN}) where {T, SGN}
+    M = zeros(T, J.n, J.n)
+    for i = 1:2:J.n-1
+        M[i+1,i] = -SGN
+        M[i,i+1] = SGN
     end
     return M
 end
 
 Base.Array(A::JMatrix) = Matrix(A)
 
-function SkewHermTridiagonal(J::JMatrix{T}) where {T}
-    vec = zeros(T, J.n - 1)
-    for i = 1 : 2 : J.n - 1
-        vec[i] = -1
-    end
-    return SkewHermTridiagonal(vec)
+function SkewHermTridiagonal(J::JMatrix{T, SGN}) where {T, SGN}
+    ev = zeros(T, J.n-1)
+    ev[1:2:end] .= -SGN
+    return SkewHermTridiagonal(ev)
 end
 
-Base.@propagate_inbounds function Base.getindex(J::JMatrix{T}, i::Integer, j::Integer) where T
-
+Base.@propagate_inbounds function Base.getindex(J::JMatrix{T, SGN}, i::Integer, j::Integer) where {T, SGN}
     @boundscheck checkbounds(J, i, j)
-    if i == j + 1 && i%2 == 0
-        return -J.sgn
-    elseif i + 1 == j && j%2 == 0
-        return J.sgn
+    if i == j + 1 && iseven(i)
+        return T(-SGN)
+    elseif i + 1 == j && iseven(j)
+        return T(SGN)
     else
         return zero(T)
     end
 end
 
-function Base.:*(J::JMatrix,A::StridedVecOrMat)
+function Base.:*(J::JMatrix{T,SGN}, A::StridedVecOrMat) where {T,SGN}
+    LA.require_one_based_indexing(A)
     m, k = size(A, 1), size(A, 2)
-    if !(m == J.n)
+    if m != J.n
         throw(DimensionMismatch("J has second dimension $(size(J,2)), A has first dimension $(size(A,1))"))
     end
-    B = similar(A, J.n, k)
-    for i = 1 : 2 : J.n-1
-        B[i,:] .= A[i+1,:].*J.sgn
-        B[i+1,:] .= -A[i,:].*J.sgn
```

Comment:
Not that `A[i,:]` was allocating here.  Even if you add `@views` to make this a view, it is probably still better to write the explicit loop (e.g. to [access the arrays in memory order](https://docs.julialang.org/en/v1/manual/performance-tips/#man-performance-column-major)).